elasticsearch/helpers/vectorstore/_async/strategies.py

```diff
@@ -0,0 +1,452 @@
+#
+# 	http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing,
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Optional, Tuple, Union, cast
+
+from elasticsearch import AsyncElasticsearch
+from elasticsearch.helpers.vectorstore._async._utils import model_must_be_deployed
+from elasticsearch.helpers.vectorstore._utils import DistanceMetric
+
+
+class AsyncRetrievalStrategy(ABC):
+    @abstractmethod
+    def es_query(
+        self,
+        query: Optional[str],
+        query_vector: Optional[List[float]],
+        text_field: str,
+        vector_field: str,
+        k: int,
+        num_candidates: int,
+        filter: List[Dict[str, Any]] = [],
+    ) -> Dict[str, Any]:
+        """
+        Returns the Elasticsearch query body for the given parameters.
+        The store will execute the query.
+
+        Args:
+            query: The text query. Can be None if query_vector is given.
+            k: The total number of results to retrieve.
+            num_candidates: The number of results to fetch initially in knn search.
+            filter: List of filter clauses to apply to the query.
+            query_vector: The query vector. Can be None if a query string is given.
+
+        Returns:
+            Dict: The Elasticsearch query body.
+        """
+
+    @abstractmethod
+    def es_mappings_settings(
+        self,
+        text_field: str,
+        vector_field: str,
+        num_dimensions: Optional[int],
+    ) -> Tuple[Dict[str, Any], Dict[str, Any]]:
+        """
+        Create the required index and do necessary preliminary work, like
+        creating inference pipelines or checking if a required model was deployed.
+
+        Args:
+            client: Elasticsearch client connection.
+            index_name: The name of the Elasticsearch index to create.
+            metadata_mapping: Flat dictionary with field and field type pairs that
+                describe the schema of the metadata.
+        """
+
+    async def before_index_creation(
+        self, client: AsyncElasticsearch, text_field: str, vector_field: str
+    ) -> None:
+        """
+        Executes before the index is created. Used for setting up
+        any required Elasticsearch resources like a pipeline.
+
+        Args:
+            client: The Elasticsearch client.
+            text_field: The field containing the text data in the index.
+            vector_field: The field containing the vector representations in the index.
+        """
+        pass
+
+    def needs_inference(self) -> bool:
+        """
+        TODO
+        """
+        return False
+
+
+class AsyncSparseVector(AsyncRetrievalStrategy):
```

Comment:
I'm wondering if there is any risk of naming collisions for these strategies. The names `DenseVector`, `SparseVector`, `BM25` seem pretty vague and likely to represent other things that are not strategies. Any reason why we are not using a `Stragegy` suffix on all these classes?

Reply:
Mainly because aesthetically I didn't want to have the Java-esque `AsyncDenseVectorScriptScoreStrategy` 😂 
But I get your point and it's probably the right thing to do.